autopilot/purepursuitwaypointfollower.cpp

```diff
@@ -90,11 +91,51 @@ void PurepursuitWaypointFollower::resetState()
     mCurrentState.currentWaypointIndex = mWaypointList.size();
 }
 
+
+double PurepursuitWaypointFollower::getCurvatureWithTrailer(QSharedPointer<TruckState> truckState, const QPointF &point, PosType vehiclePosType)
+{
+    const PosPoint vehiclePos = truckState->getPosition(vehiclePosType);
+    double currYaw_rad = vehiclePos.getYaw() * M_PI / 180.0;
+    double measuredTrailerAngle = truckState->getTrailerAngleRadians() ;
+    double l2 = 0.715;; // trailer wheelbase in meters 
+
+    if (truckState->getSpeed()> 0 ){
+        QPointF pointInVehicleFrame;
+        pointInVehicleFrame.setX(point.x()-vehiclePos.getX());
+        pointInVehicleFrame.setY(point.y()-vehiclePos.getY());
+        double theta_err =  atan2(pointInVehicleFrame.y(), pointInVehicleFrame.x()) - currYaw_rad; 
+        double desired_hitch_angle = atan(2*l2*sin(theta_err)); // desired trailer/hitch angle
+        double k=1; //gain
+        return  k*( measuredTrailerAngle - desired_hitch_angle ) - ( sin(measuredTrailerAngle)/ (l2)) ;
+
+    } else {
+        double trailerYaw = currYaw_rad - measuredTrailerAngle ; 
+        double delta_x_ = (l2) * cos( trailerYaw); 
+        double delta_y = (l2) * sin( trailerYaw);
+        double trailerPositionx = vehiclePos.getX() - delta_x_;
+        double trailerPositiony = vehiclePos.getY() - delta_y; 
+        
+        QPointF pointInTrailerFrame; 
+        pointInTrailerFrame.setX(point.x()-trailerPositionx); 
+        pointInTrailerFrame.setY(point.y()-trailerPositiony); 
+        double theta_err =  atan2(pointInTrailerFrame.y(), pointInTrailerFrame.x()) - trailerYaw;        
+        double desired_hitch_angle = atan(2*l2*sin(theta_err) / truckState->getAutopilotRadius() );
+        double k=-2.5; // gain
+        return  k*( measuredTrailerAngle - desired_hitch_angle ) - ( sin(measuredTrailerAngle)/ (l2)) ;
+    }
+}
+
 double PurepursuitWaypointFollower::getCurvatureToPointInENU(const QPointF &point)
 {
     QSharedPointer<VehicleState> vehicleState = isOnVehicle() ? mMovementController->getVehicleState() : mVehicleConnection->getVehicleState();
-
+    
+    auto truckState = qSharedPointerDynamicCast<TruckState>(vehicleState);
+    if (truckState){
+        return getCurvatureWithTrailer(truckState, point, mPosTypeUsed);
+    }
+    
     return getCurvatureToPointInVehicleFrame(coordinateTransforms::ENUToVehicleFrame(point, vehicleState->getPosition(mPosTypeUsed).getXYZ()));
+
 }
```

Comment:
I think we need to move getCurvatureWithTrailer() out of purepursuitwaypointfollower. It looks like getCurvatureWithTrailer() is a version of PurepursuitWaypointFollower::getCurvatureToPointInVehicleFrame(), which we should move to a more general place. I propose we move it to vehiclestate and make it virtual, this way we can move getCurvatureWithTrailer() to the truckstate, and the current implementation of PurepursuitWaypointFollower::getCurvatureToPointInVehicleFrame() to the carstate. It would be something like this

VehicleState::getCurvatureToPointInVehicleFrame() // virtual
TruckState::getCurvatureToPointInVehicleFrame() // override
CarState::getCurvatureToPointInVehicleFrame() // override

This way, we can call the function using only vehiclestate like

```suggestion
double PurepursuitWaypointFollower::getCurvatureToPointInENU(const QPointF &point)
{
    QSharedPointer<VehicleState> vehicleState = isOnVehicle() ? mMovementController->getVehicleState() : mVehicleConnection->getVehicleState();

    return vehicleState->getCurvatureToPointInVehicleFrame(coordinateTransforms::ENUToVehicleFrame(point, vehicleState->getPosition(mPosTypeUsed).getXYZ()));
}
```
All following commits can be removed and we dont clutter purepursuitwaypointfollower with variables connected to a specific vehicle.

Reply:
I agree with this. I think it's also more cleaner this way than my earlier suggestion. I will revert the commits to https://github.com/RISE-Dependable-Transport-Systems/WayWise/pull/103/commits/40c78658bb3bf68edea81b5cdedf26003a61ee5e. They are available [here ](https://github.com/RISE-Dependable-Transport-Systems/WayWise/tree/temp/pure-pursuit-with-a-trailer-var) for reference.

@chrpro: As it was your original work, please incorporate these suggestions or let me know otherwise. It would be useful to bring these changes to the main branch so they can be used in WayWiseR and I would be glad to help if needed.